dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,580 @@
+{
+    assert(frame_end.block == partition_end.block
+           || frame_end.block + 1 == partition_end.block);
+
+    // If window only have row_number or rank/dense_rank functions, set frame_end to the next row of current_row and frame_ended to true
+    if (only_have_pure_window)
+    {
+        frame_end = current_row;
+        advanceRowNumber(frame_end);
+        frame_ended = true;
+        return;
+    }
+
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                    "window function only support pure window function now.");
+}
+
+void WindowBlockInputStream::advanceFrameEnd()
+{
+    // frame_end must be greater or equal than frame_start, so if the
+    // frame_start is already past the current frame_end, we can start
+    // from it to save us some work.
+    if (frame_end < frame_start)
+    {
+        frame_end = frame_start;
+    }
+
+    // No reason for this function to be called again after it succeeded.
+    assert(!frame_ended);
+
+    // switch for another frame type
+    switch (window_description.frame.end_type)
+    {
+    case WindowFrame::BoundaryType::Current:
+        advanceFrameEndCurrentRow();
+        break;
+    case WindowFrame::BoundaryType::Unbounded:
+    case WindowFrame::BoundaryType::Offset:
+    default:
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                        "The frame end type '{}' is not implemented",
+                        window_description.frame.end_type);
+    }
+}
+
+void WindowBlockInputStream::writeOutCurrentRow()
+{
+    assert(current_row < partition_end);
+    assert(current_row.block >= first_block_number);
+
+    for (size_t wi = 0; wi < workspaces.size(); ++wi)
+    {
+        auto & ws = workspaces[wi];
+        ws.window_function->windowInsertResultInto(this->shared_from_this(), wi);
+    }
+}
+
+Block WindowBlockInputStream::tryGetOutputBlock()
+{
+    assert(first_not_ready_row.block >= first_block_number);
+    // The first_not_ready_row might be past-the-end if we have already
+    // calculated the window functions for all input rows. That's why the
+    // equality is also valid here.
+    assert(first_not_ready_row.block <= first_block_number + window_blocks.size());
+    assert(next_output_block_number >= first_block_number);
+
+    if (next_output_block_number < first_not_ready_row.block)
+    {
+        const auto i = next_output_block_number - first_block_number;
+        auto & block = window_blocks[i];
+        auto columns = block.input_columns;
+        for (auto & res : block.output_columns)
+        {
+            columns.push_back(ColumnPtr(std::move(res)));
+        }
+        ++next_output_block_number;
+
+        auto output_block = output_header.cloneWithColumns(std::move(columns));
+        releaseAlreadyOutputWindowBlock();
+        return output_block;
+    }
+    return {};
+}
+
+bool WindowBlockInputStream::onlyHaveRowNumber()
+{
+    return workspaces.size() == 1 && workspaces[0].window_function->getName() == "row_number";
```

Comment:
What if there are multiple window functions, all of them are `row_number`?

Reply:
yes, done.